File: ios/Classes/NativeFeedPlayerPlugin.swift
import Flutter
import Foundation
import UIKit

/// Holds the Pigeon sink for one event channel and buffers events emitted
/// before Dart subscribes.
///
/// Native playback starts reporting state during `createController`, which runs
/// before the caller has had a chance to listen to that controller's stream, so
/// without a small buffer the first transition is silently lost.
final class BufferedEventSink<T> {
  private let maxBuffered: Int
  private var pending: [T] = []
  private var sink: PigeonEventSink<T>?

  init(maxBuffered: Int = 64) {
    self.maxBuffered = maxBuffered
  }

  func attach(_ sink: PigeonEventSink<T>) {
    self.sink = sink
    let buffered = pending
    pending.removeAll()
    for event in buffered {
      sink.success(event)
    }
  }

  func detach() {
    sink = nil
    pending.removeAll()
  }

  func emit(_ event: T) {
    if let sink {
      sink.success(event)
      return
    }
    if pending.count >= maxBuffered {
      pending.removeFirst()
    }
    pending.append(event)
  }
}

/// Pigeon generates a distinct handler class per event channel, so each one
/// needs a concrete adapter that forwards its sink into a shared holder.
private final class PlaybackStateStreamAdapter: PlaybackStateEventsStreamHandler {
  private let holder: BufferedEventSink<PlaybackStateEvent>

  init(holder: BufferedEventSink<PlaybackStateEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<PlaybackStateEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class PositionStreamAdapter: PositionEventsStreamHandler {
  private let holder: BufferedEventSink<PositionEvent>

  init(holder: BufferedEventSink<PositionEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<PositionEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class MetricsStreamAdapter: MetricsEventsStreamHandler {
  private let holder: BufferedEventSink<MetricsEvent>

  init(holder: BufferedEventSink<MetricsEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<MetricsEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class LifecycleStreamAdapter: LifecycleEventsStreamHandler {
  private let holder: BufferedEventSink<ControllerLifecycleEvent>

  init(holder: BufferedEventSink<ControllerLifecycleEvent>) {
    self.holder = holder
  }

  override func onListen(
    withArguments arguments: Any?,
    sink: PigeonEventSink<ControllerLifecycleEvent>
  ) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

public final class NativeFeedPlayerPlugin: NSObject, FlutterPlugin, NativeFeedPlayerHostApi {
  private static let videoViewType = "native_feed_player/video_view"

  /// Controller ids must stay unique for the life of the process, not the life
  /// of one engine attachment, so a re-attaching engine cannot mint ids that
  /// collide with handles Dart still holds.
  private static var controllerIdSeed = 0

  private let stateEvents = BufferedEventSink<PlaybackStateEvent>()
  private let positionEvents = BufferedEventSink<PositionEvent>()
  private let metricsEvents = BufferedEventSink<MetricsEvent>()
  private let lifecycleEvents = BufferedEventSink<ControllerLifecycleEvent>()

  private var manager: AVPlayerManager?
  private var memoryWarningObserver: NSObjectProtocol?
  private let renderViewPool = RenderViewPool(maxPoolSize: 8)
  private var videoViews: [Int64: NativeVideoPlatformView] = [:]
  private var attachedControllerByViewId: [Int64: Int] = [:]
  private var binaryMessenger: FlutterBinaryMessenger?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = NativeFeedPlayerPlugin()
    instance.binaryMessenger = registrar.messenger()
    instance.configure(messenger: registrar.messenger())

    let viewFactory = NativeVideoViewFactory(
      renderViewPool: instance.renderViewPool,
      onCreate: { [weak instance] viewId, view in
        instance?.videoViews[viewId] = view
      },
      onDispose: { [weak instance] viewId, renderView in
        instance?.handleVideoViewDisposed(viewId: viewId, renderView: renderView)
      }
    )
    registrar.register(viewFactory, withId: videoViewType)
    NativeFeedPlayerHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
  }

  deinit {
    if let observer = memoryWarningObserver {
      NotificationCenter.default.removeObserver(observer)
      memoryWarningObserver = nil
    }
    if let binaryMessenger {
      NativeFeedPlayerHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: nil)
    }
    videoViews.removeAll()
    attachedControllerByViewId.removeAll()
    renderViewPool.clear()
    manager?.disposeAll()
  }

  // MARK: - Host API

  func initialize(config: FeedPlayerConfigMessage) throws {
    try managerOrThrow().initialize(config: config)
  }

  func setSources(sources: [FeedSourceMessage]) throws {
    try managerOrThrow().setSources(sources.map { $0.toRegisteredSource() })
  }

  func appendSources(sources: [FeedSourceMessage]) throws {
    try managerOrThrow().appendSources(sources.map { $0.toRegisteredSource() })
  }

  func removeSources(request: SourceIdsRequest) throws {
    try managerOrThrow().removeSources(request.sourceIds)
  }

  func createController(request: CreateControllerRequest) throws -> Int64 {
    if request.sourceId.isEmpty {
      throw PigeonError(
        code: "invalid_source",
        message: "createController requires a source id.",
        details: nil
      )
    }

    NativeFeedPlayerPlugin.controllerIdSeed += 1
    let controllerId = NativeFeedPlayerPlugin.controllerIdSeed

    do {
      try managerOrThrow().createController(
        controllerId: controllerId,
        sourceId: request.sourceId,
        autoPlay: request.autoPlay,
        looping: request.looping
      )
    } catch let setupError as AVPlayerManager.PlaybackSetupError {
      throw PigeonError(code: setupError.code, message: setupError.message, details: nil)
    } catch {
      throw PigeonError(
        code: "create_failed",
        message: error.localizedDescription,
        details: nil
      )
    }

    return Int64(controllerId)
  }

  func disposeController(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().disposeController(controllerId: controllerId)
    }
  }

  func play(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().play(controllerId: controllerId)
    }
  }

  func pause(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().pause(controllerId: controllerId)
    }
  }

  func seekTo(request: SeekRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().seekTo(controllerId: controllerId, positionMs: request.positionMs)
    }
  }

  func setVisibleSource(request: VisibleSourceRequest) throws {
    try managerOrThrow().setVisibleSource(request.sourceId)
  }

  func evictCachedMedia(request: SourceIdsRequest) throws {
    try managerOrThrow().evictCachedMedia(request.sourceIds)
  }

  func clearMediaCache() throws {
    try managerOrThrow().clearMediaCache()
  }

  func cacheStatus(request: VisibleSourceRequest) throws -> CacheStatusMessage {
    try managerOrThrow().cacheStatus(sourceId: request.sourceId)
  }

  func cacheUsageBytes() throws -> Int64 {
    try managerOrThrow().cacheUsageBytes()
  }

  func attachView(request: AttachViewRequest) throws {
    let controllerId = Int(request.controllerId)
    let viewId = Int64(request.viewId)
    guard controllerId > 0, viewId >= 0 else {
      throw PigeonError(
        code: "invalid_attach",
        message: "attachView requires valid controllerId and viewId.",
        details: nil
      )
    }
    guard let platformView = videoViews[viewId] else {
      throw PigeonError(
        code: "view_not_found",
        message: "No video view found for id=\(viewId).",
        details: nil
      )
    }

    let manager = try managerOrThrow()
    if let previous = attachedControllerByViewId[viewId], previous != controllerId {
      manager.detach(controllerId: previous)
    }
    attachedControllerByViewId[viewId] = controllerId
    manager.attach(controllerId: controllerId, renderView: platformView.renderView)
  }

  func detachView(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    guard controllerId > 0 else {
      return
    }
    let manager = try managerOrThrow()
    manager.detach(controllerId: controllerId)
    for (viewId, attached) in attachedControllerByViewId where attached == controllerId {
      attachedControllerByViewId.removeValue(forKey: viewId)
    }
  }

  func disposeAll() throws {
    try managerOrThrow().disposeAll()
    attachedControllerByViewId.removeAll()
  }

  // MARK: - Wiring

  private func configure(messenger: FlutterBinaryMessenger) {
    PlaybackStateEventsStreamHandler.register(
      with: messenger,
      streamHandler: PlaybackStateStreamAdapter(holder: stateEvents)
    )
    PositionEventsStreamHandler.register(
      with: messenger,
      streamHandler: PositionStreamAdapter(holder: positionEvents)
    )
    MetricsEventsStreamHandler.register(
      with: messenger,
      streamHandler: MetricsStreamAdapter(holder: metricsEvents)
    )
    LifecycleEventsStreamHandler.register(
      with: messenger,
      streamHandler: LifecycleStreamAdapter(holder: lifecycleEvents)
    )

    manager = AVPlayerManager(
      onState: { [weak self] controllerId, status, error in
        self?.onMain {
          self?.stateEvents.emit(
            PlaybackStateEvent(
              controllerId: Int64(controllerId),
              status: status,
              error: error
            )
          )
        }
      },
      onReleased: { [weak self] controllerId, reason in
        self?.onMain {
          self?.lifecycleEvents.emit(
            ControllerLifecycleEvent(controllerId: Int64(controllerId), reason: reason)
          )
        }
      },
      onPosition: { [weak self] event in
        self?.onMain { self?.positionEvents.emit(event) }
      },
      onMetrics: { [weak self] event in
        self?.onMain { self?.metricsEvents.emit(event) }
      }
    )

    if let observer = memoryWarningObserver {
      NotificationCenter.default.removeObserver(observer)
      memoryWarningObserver = nil
    }
    memoryWarningObserver = NotificationCenter.default.addObserver(
      forName: UIApplication.didReceiveMemoryWarningNotification,
      object: nil,
      queue: .main
    ) { [weak self] _ in
      self?.manager?.onMemoryWarning()
    }
  }

  private func onMain(_ work: @escaping () -> Void) {
    if Thread.isMainThread {
      work()
    } else {
      DispatchQueue.main.async(execute: work)
    }
  }

  private func handleVideoViewDisposed(viewId: Int64, renderView: NativeVideoRenderView) {
    if let controllerId = attachedControllerByViewId.removeValue(forKey: viewId) {
      manager?.detach(controllerId: controllerId)
    }
    videoViews.removeValue(forKey: viewId)
    renderViewPool.release(renderView)
  }

  private func managerOrThrow() throws -> AVPlayerManager {
    guard let manager else {
      throw PigeonError(
        code: "not_attached",
        message: "NativeFeedPlayerPlugin is not attached to a Flutter engine.",
        details: nil
      )
    }
    return manager
  }
}

extension FeedSourceMessage {
  fileprivate func toRegisteredSource() -> RegisteredSource {
    RegisteredSource(
      id: id,
      uri: uri,
      rank: Int(rank),
      kind: kind,
      headers: headers
    )
  }
}

File: android/src/main/kotlin/com/example/native_feed_player/NativeFeedPlayerPlugin.kt
package com.example.native_feed_player

import android.content.ComponentCallbacks2
import android.content.Context
import android.content.res.Configuration
import android.view.TextureView
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import java.util.concurrent.atomic.AtomicInteger

/** NativeFeedPlayerPlugin */
class NativeFeedPlayerPlugin : FlutterPlugin, ComponentCallbacks2, NativeFeedPlayerHostApi {
    private companion object {
        private const val VIDEO_VIEW_TYPE = "native_feed_player/video_view"

        /**
         * Controller ids must stay unique for the life of the process, not the
         * life of one engine attachment. Restarting the counter on detach lets
         * a second engine (or a hot restart) mint ids that collide with handles
         * Dart still holds.
         */
        private val controllerIdSeed = AtomicInteger(0)
    }

    private var appContext: Context? = null
    private var exoPlayerManager: ExoPlayerManager? = null
    private var binaryMessenger: BinaryMessenger? = null

    private val stateEvents = BufferedStreamHandler<PlaybackStateEvent>()
    private val positionEvents = BufferedStreamHandler<PositionEvent>()
    private val metricsEvents = BufferedStreamHandler<MetricsEvent>()
    private val lifecycleEvents = BufferedStreamHandler<ControllerLifecycleEvent>()

    private val textureViewPool = TextureViewPool(maxPoolSize = 8)
    private val videoViews = mutableMapOf<Int, NativeVideoPlatformView>()
    private val attachedControllerByViewId = mutableMapOf<Int, Int>()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        appContext = flutterPluginBinding.applicationContext
        appContext?.registerComponentCallbacks(this)
        binaryMessenger = flutterPluginBinding.binaryMessenger

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            VIDEO_VIEW_TYPE,
            NativeVideoViewFactory(
                textureViewPool = textureViewPool,
                onCreate = { viewId, view -> videoViews[viewId] = view },
                onDispose = { viewId, textureView -> handleVideoViewDisposed(viewId, textureView) }
            )
        )

        PlaybackStateEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            PlaybackStateStreamAdapter(stateEvents)
        )
        PositionEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            PositionStreamAdapter(positionEvents)
        )
        MetricsEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            MetricsStreamAdapter(metricsEvents)
        )
        LifecycleEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            LifecycleStreamAdapter(lifecycleEvents)
        )

        exoPlayerManager = ExoPlayerManager(
            context = flutterPluginBinding.applicationContext,
            onState = { controllerId, status, error ->
                stateEvents.emit(
                    PlaybackStateEvent(
                        controllerId = controllerId.toLong(),
                        status = status,
                        error = error
                    )
                )
            },
            onReleased = { controllerId, reason ->
                lifecycleEvents.emit(
                    ControllerLifecycleEvent(
                        controllerId = controllerId.toLong(),
                        reason = reason
                    )
                )
            },
            onPosition = { event -> positionEvents.emit(event) },
            onMetrics = { event -> metricsEvents.emit(event) }
        )

        NativeFeedPlayerHostApi.setUp(
            binaryMessenger = flutterPluginBinding.binaryMessenger,
            api = this
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext?.unregisterComponentCallbacks(this)
        appContext = null
        exoPlayerManager?.disposeAll()
        exoPlayerManager = null
        videoViews.clear()
        attachedControllerByViewId.clear()
        textureViewPool.clear()
        stateEvents.detach()
        positionEvents.detach()
        metricsEvents.detach()
        lifecycleEvents.detach()
        NativeFeedPlayerHostApi.setUp(binaryMessenger = binding.binaryMessenger, api = null)
        binaryMessenger = null
    }

    override fun initialize(config: FeedPlayerConfigMessage) {
        managerOrThrow().initialize(config)
    }

    override fun setSources(sources: List<FeedSourceMessage>) {
        managerOrThrow().setSources(sources.map { it.toRegisteredSource() })
    }

    override fun appendSources(sources: List<FeedSourceMessage>) {
        managerOrThrow().appendSources(sources.map { it.toRegisteredSource() })
    }

    override fun removeSources(request: SourceIdsRequest) {
        managerOrThrow().removeSources(request.sourceIds)
    }

    override fun createController(request: CreateControllerRequest): Long {
        if (request.sourceId.isBlank()) {
            throw FlutterError("invalid_source", "createController requires a source id.", null)
        }
        val controllerId = controllerIdSeed.incrementAndGet()
        managerOrThrow().createController(
            controllerId = controllerId,
            sourceId = request.sourceId,
            autoPlay = request.autoPlay,
            looping = request.looping
        )
        return controllerId.toLong()
    }

    override fun disposeController(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().disposeController(controllerId)
        }
    }

    override fun play(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().play(controllerId)
        }
    }

    override fun pause(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().pause(controllerId)
        }
    }

    override fun seekTo(request: SeekRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().seekTo(controllerId, request.positionMs)
        }
    }

    override fun setVisibleSource(request: VisibleSourceRequest) {
        managerOrThrow().setVisibleSource(request.sourceId)
    }

    override fun evictCachedMedia(request: SourceIdsRequest) {
        // Wired to the disk cache in a later phase; preload state is dropped so
        // the next request re-fetches.
        managerOrThrow().removeSources(emptyList())
    }

    override fun clearMediaCache() {
        // No persistent cache yet; see docs/ARCHITECTURE.md for the roadmap.
    }

    override fun cacheStatus(request: VisibleSourceRequest): CacheStatusMessage {
        return CacheStatusMessage(
            sourceId = request.sourceId,
            cachedBytes = 0,
            totalBytes = 0,
            isComplete = false
        )
    }

    override fun cacheUsageBytes(): Long = 0

    override fun attachView(request: AttachViewRequest) {
        val controllerId = request.controllerId.toInt()
        val viewId = request.viewId.toInt()
        if (controllerId <= 0 || viewId < 0) {
            throw FlutterError(
                "invalid_attach",
                "attachView requires valid controllerId and viewId.",
                null
            )
        }

        val view = videoViews[viewId]
            ?: throw FlutterError("view_not_found", "No video view found for id=$viewId.", null)

        val manager = managerOrThrow()
        val previousController = attachedControllerByViewId[viewId]
        if (previousController != null && previousController != controllerId) {
            manager.detachControllerFromView(previousController)
        }
        attachedControllerByViewId[viewId] = controllerId
        manager.attachControllerToView(controllerId, view.textureView)
    }

    override fun detachView(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId <= 0) {
            return
        }
        val manager = managerOrThrow()
        manager.detachControllerFromView(controllerId)
        val staleViewIds = attachedControllerByViewId
            .filterValues { it == controllerId }
            .keys
            .toList()
        for (viewId in staleViewIds) {
            attachedControllerByViewId.remove(viewId)
        }
    }

    override fun disposeAll() {
        managerOrThrow().disposeAll()
        attachedControllerByViewId.clear()
    }

    override fun onTrimMemory(level: Int) {
        exoPlayerManager?.onTrimMemory(level)
    }

    @Deprecated("ComponentCallbacks2.onLowMemory is deprecated by the platform")
    override fun onLowMemory() {
        exoPlayerManager?.onLowMemory()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        // No-op: handled by the Flutter engine and ExoPlayer internally.
    }

    private fun managerOrThrow(): ExoPlayerManager {
        return exoPlayerManager
            ?: throw FlutterError(
                "not_attached",
                "NativeFeedPlayerPlugin is not attached to a Flutter engine.",
                null
            )
    }

    private fun handleVideoViewDisposed(viewId: Int, textureView: TextureView) {
        val controllerId = attachedControllerByViewId.remove(viewId)
        if (controllerId != null) {
            exoPlayerManager?.detachControllerFromView(controllerId)
        }
        videoViews.remove(viewId)
        textureViewPool.release(textureView)
    }
}

private fun FeedSourceMessage.toRegisteredSource(): RegisteredSource = RegisteredSource(
    id = id,
    uri = uri,
    rank = rank.toInt(),
    kind = kind,
    headers = headers
)
